### PR TITLE
8281296: Create a regression test for JDK-4515999

### PR DESCRIPTION
**Pull request by manukumarvs**
Create a regression test for [JDK-4515999](https://bugs.openjdk.java.net/browse/JDK-4515999)
Create a test which checks whether incrementing dates via the keyboard (up/down) gives the same results as when using mouse press on the arrow buttons in a JSpinner.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281296](https://bugs.openjdk.java.net/browse/JDK-8281296): Create a regression test for JDK-4515999

### Reviewers
 * [Alexey Ivanov](https://openjdk.java.net/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7363/head:pull/7363` \
`$ git checkout pull/7363`

Update a local copy of the PR: \
`$ git checkout pull/7363` \
`$ git pull https://git.openjdk.java.net/jdk pull/7363/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7363`

View PR using the GUI difftool: \
`$ git pr show -t 7363`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7363.diff">https://git.openjdk.java.net/jdk/pull/7363.diff</a>

</details>

### CODE REVIEW

**Comment by manukumarvs** [comment]
@aivanov-jdk Can you please review this?

File: test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+ * @key headful
+ * @bug 4515999
+ * @summary Check whether incrementing dates via the keyboard (up/down) gives
+ * the same results as using mouse press on the arrow buttons in a JSpinner.
+ * @run main JSpinnerMouseAndKeyPressTest
+ */
+public class JSpinnerMouseAndKeyPressTest {
+    //2 days in milliseconds
+    private static final int EXPECTED_VALUE_2_DAYS = 2 * 24 * 60 * 60 * 1000;
+    private static JFrame frame;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java:32]: I suggest adding a blank line between the constant and the frame to separate them.

File: test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+/*
+ * @test
+ * @key headful
+ * @bug 4515999
+ * @summary Check whether incrementing dates via the keyboard (up/down) gives
+ * the same results as using mouse press on the arrow buttons in a JSpinner.
+ * @run main JSpinnerMouseAndKeyPressTest
+ */
+public class JSpinnerMouseAndKeyPressTest {
+    //2 days in milliseconds
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java:30]: Please add a space after `//` according to [the guidelines](http://cr.openjdk.java.net/~alundblad/styleguide/index-v6.html#toc-whitespace). In other places too.

File: test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+
+    public static void runTest() throws Exception {
+        Robot robot = new Robot();
+        robot.setAutoWaitForIdle(true);
+        robot.setAutoDelay(100);
+        List<String> lafs = Arrays.stream(getInstalledLookAndFeels())
+                                  .map(UIManager.LookAndFeelInfo::getClassName)
+                                  .collect(Collectors.toList());
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java:72]: This is probably a bit too much for just mapping `LookAndFeelInfo` to `lafClassName`… If you like it though, I don't mind.

File: test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            SwingUtilities.invokeAndWait(() -> {
+                setLookAndFeel(laf);
+                createUI();
+            });
+            try {
+                final Point spinnerLocationOnScreen = spinner.getLocationOnScreen();
+                final int spinnerEditorWidth = spinner.getEditor().getWidth();
+                final int spinnerButtonWidth = spinner.getWidth() -
+                        spinnerEditorWidth;
+                final int spinnerQuarterHeight = spinner.getHeight() / 4;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java:83]: The `spinner` component must be accessed from EDT only.

File: test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                spinnerDownButtonCenter.y = spinnerLocationOnScreen.y +
+                        3 * spinnerQuarterHeight;
+
+                //Mouse press use-case
+                // Move Mouse pointer to UP button center and click it
+                robot.mouseMove(spinnerUpButtonCenter.x, spinnerUpButtonCenter.y);
+                robot.mousePress(InputEvent.BUTTON1_MASK);
+                robot.mouseRelease(InputEvent.BUTTON1_MASK);
+
+                long upValue = ((Date) spinner.getValue()).getTime();
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java:102]: The `spinner` component must be accessed from EDT only.

File: test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                Point spinnerDownButtonCenter = new Point();
+                spinnerDownButtonCenter.x = spinnerUpButtonCenter.x;
+                spinnerDownButtonCenter.y = spinnerLocationOnScreen.y +
+                        3 * spinnerQuarterHeight;
+
+                //Mouse press use-case
+                // Move Mouse pointer to UP button center and click it
+                robot.mouseMove(spinnerUpButtonCenter.x, spinnerUpButtonCenter.y);
+                robot.mousePress(InputEvent.BUTTON1_MASK);
+                robot.mouseRelease(InputEvent.BUTTON1_MASK);
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java:100]: `BUTTON1_MASK` is deprecated in favour of `BUTTON1_DOWN_MASK`

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]
Does this new test reproduce the original problem reported in JDK-4515999?

File: test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                final int spinnerEditorWidth = spinner.getEditor().getWidth();
+                final int spinnerButtonWidth = spinner.getWidth() -
+                        spinnerEditorWidth;
+                final int spinnerQuarterHeight = spinner.getHeight() / 4;
+
+                Point spinnerUpButtonCenter = new Point();
+                spinnerUpButtonCenter.x = spinnerLocationOnScreen.x + spinnerEditorWidth +
+                        (spinnerButtonWidth / 2);
+                spinnerUpButtonCenter.y = spinnerLocationOnScreen.y +
+                        spinnerQuarterHeight;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JSpinner/4515999/JSpinnerMouseAndKeyPressTest.java:89]: The [Java Style Guidelines](http://cr.openjdk.java.net/~alundblad/styleguide/index-v6.html#toc-wrapping-lines) suggests breaking before the operators.

**Comment by manukumarvs** [comment]
> Does this new test reproduce the original problem reported in JDK-4515999?

Yes it's reproducible in JDK 1.4.

**Review by aivanov-jdk — APPROVED** [review]

**Pull request closed without merging**